webapp/graphite/util.py

```diff
@@ -220,48 +221,49 @@ def write_index(whisper_dir=None, ceres_dir=None, index=None):
   return None
 
 
-def logtime(custom_msg=False, custom_name=False):
-  def wrap(f):
-    def wrapped_f(*args, **kwargs):
-      msg = 'completed in'
-      name = f.__module__ + '.' + f.__name__
-
-      t = time.time()
-      if custom_msg:
-        def set_msg(msg):
-          wrapped_f.msg = msg
-
-        kwargs['msg_setter'] = set_msg
-      if custom_name:
-        def set_name(name):
-          wrapped_f.name = name
-
-        kwargs['name_setter'] = set_name
-
-      try:
-        res = f(*args, **kwargs)
-      except:
-        msg = 'failed in'
-        raise
-      finally:
-        msg = getattr(wrapped_f, 'msg', msg)
-        name = getattr(wrapped_f, 'name', name)
-
-        log.info(
-          '{name} :: {msg} {sec:.6}s'.format(
-            name=name,
-            msg=msg,
-            sec=time.time() - t,
-          )
-        )
-
-      return res
-    return wrapped_f
-  return wrap
-
-
-@logtime(custom_msg=True)
-def build_index(base_path, extension, fd, msg_setter=None):
+class Timer(object):
+  __slots__ = ('msg', 'name', 'start_time')
+
+  def __init__(self, name):
+    self.name = name
+    self.msg = 'completed in'
+    self.start_time = time.time()
+
+  def set_msg(self, msg):
+    self.msg = msg
+
+  def set_name(self, name):
+    self.name = name
+
+  def stop(self):
+    log.info(
+      '{name} :: {msg} {sec:.6}s'.format(
+        name=self.name,
+        msg=self.msg,
+        sec=time.time() - self.start_time,
+      )
+    )
+
+
+def logtime(f):
+  @wraps(f)
+  def wrapped_f(*args, **kwargs):
+    timer = Timer(f.__module__ + '.' + f.__name__)
+    kwargs['timer'] = timer
+
+    try:
+      return f(*args, **kwargs)
+    except:
+      timer.msg = 'failed in'
+      raise
+    finally:
+      timer.stop()
+
+  return wrapped_f
+
+
+@logtime
+def build_index(base_path, extension, fd, timer=None):
```

Comment:
This refactor simplifies use of `@logtime`, and makes it more extensible in future by adding functionality to the `Timer` class.